### PR TITLE
expand dimension in non sparse case as well

### PR DESCRIPTION
**Pull request by dianaml0**
## What does this PR do?
Expand dimension in non sparse case in `matmul_with_mask`

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: xformers/components/attention/core.py [hunk 1]
```diff
@@ -89,6 +89,8 @@ def _matmul_with_mask(
     # Non optimized codepath
     att = a @ b
     if mask.dtype == torch.bool:
+        if mask.ndim == 2:
```
Comment (blefaudeux) [inline xformers/components/attention/core.py:92]: it does not work automatically given the broadcasting rules ?

Comment (dianaml0) [reply xformers/components/attention/core.py:92]: Yea I was surprised broadcasting wasn't working as well, but I was getting this error: 
Screen Shot 2021-11-11 at 4 15 24 PM

Comment (blefaudeux) [reply xformers/components/attention/core.py:92]: ahh, it's because we index explicitly here, so I'm guessing that broadcasting does not apply !

**Review by blefaudeux — APPROVED** [review]
LGTM! I think that @fmassa was thinking about moving these steps out of here though, may be a good idea to keep core/ lean and clean, but it applies to the rest of it

**Comment by dianaml0** [comment]
> LGTM! I think that @fmassa was thinking about moving these steps out of here though, may be a good idea to keep core/ lean and clean, but it applies to the rest of it

That seems like a good idea. Some of the expansion happens in the `ScaledDotProductAttention` vs some happening in `_matmul_with_mask` which could be confusing

**Comment by blefaudeux** [comment]
see https://github.com/facebookresearch/xformers/issues/90, we could move all of that out of this function as @fmassa recommends in a dedicated function, and call it only in the relevant places

**Pull request merged**